### PR TITLE
[NVIDIA XLA:GPU] Fused MHA Support in XLA GPU: SE

### PR DESCRIPTION
**Pull request by AyanmoI**
cuDNN currently supports the following patterns for Multi-headed attention:

BMM1 - BMM2
BMM1 - Scale - Bias - Mask - Softmax - BMM2
BMM1 - Scale - Bias - Mask - Softmax - Dropout - BMM2
BMM1 - Scale - Mask - Softmax - BMM2
BMM1 - Scale - Mask - Softmax - Dropout - BMM2
BMM1 - Softmax - Dropout - BMM2

This PR adds support for the stream executor for these patterns.

### CODE REVIEW

**Comment by AyanmoI** [comment]
@cheshire @akuegel Opened this new PR

File: tensorflow/compiler/xla/stream_executor/dnn.cc [hunk 1]
```diff
@@ -482,6 +542,169 @@ std::string AlgorithmConfig::ToString() const {
...
+  std::vector<int64_t> minor_to_major_vec(minor_to_major.size());
+  CHECK_EQ(dimensions.size(), minor_to_major.size());
+  for (int i = 0; i < dimensions.size(); i++) {
+    dims[i] = dimensions[i];
+    minor_to_major_vec[i] = minor_to_major[i];
+  }
+  return TensorDescriptor(type, dims, minor_to_major_vec);
+}
+
+const std::string TensorDescriptor::ToString() const {
```
Comment (akuegel) [inline tensorflow/compiler/xla/stream_executor/dnn.cc:604]: Please remove the first const. ClangTidy complains about this.
Also, it wants ```#include <string>``` at the top of the file.

File: tensorflow/compiler/xla/stream_executor/dnn.cc [hunk 2]
```diff
@@ -482,6 +542,169 @@ std::string AlgorithmConfig::ToString() const {
...
+  }
+  for (int i = 0; i < contracting_dims.size(); i++) {
+    contracting_dims_vec[i] = contracting_dims[i];
+  }
+  return MatmulTensorDescriptor(
+      TensorDescriptor::For(type, dimensions, minor_to_major), batch_dims_vec,
+      contracting_dims_vec);
+}
+
+const std::string MatmulTensorDescriptor::ToString() const {
```
Comment (akuegel) [inline tensorflow/compiler/xla/stream_executor/dnn.cc:701]: Please remove the first const. ClangTidy complains about this.

File: tensorflow/compiler/xla/stream_executor/dnn.cc [hunk 3]
```diff
@@ -482,6 +542,169 @@ std::string AlgorithmConfig::ToString() const {
...
+  return absl::StrFormat(
+      "{%s, batch_dimension_numbers: %s contracting_dim: %s}",
+      tensor_.ToString(), absl::StrJoin(batch_dimension_numbers_, ","),
+      absl::StrJoin(contracting_dim_, ","));
+};
```
Comment (akuegel) [inline tensorflow/compiler/xla/stream_executor/dnn.cc:706]: Please remove the unnecessary ";".

File: tensorflow/compiler/xla/stream_executor/dnn.cc [hunk 4]
```diff
@@ -482,6 +542,169 @@ std::string AlgorithmConfig::ToString() const {
...
+    minor_to_major_vec[i] = minor_to_major[i];
+  }
+  return TensorDescriptor(type, dims, minor_to_major_vec);
+}
+
+const std::string TensorDescriptor::ToString() const {
+  return absl::StrFormat("{dimensions: %s minor_to_major: %s}",
+                         absl::StrJoin(dimensions(), ","),
+                         absl::StrJoin(minor_to_major(), ","));
+};
```
Comment (akuegel) [inline tensorflow/compiler/xla/stream_executor/dnn.cc:608]: Please remove the unnecessary ";".

File: tensorflow/compiler/xla/stream_executor/dnn.cc [hunk 5]
```diff
@@ -482,6 +542,169 @@ std::string AlgorithmConfig::ToString() const {
...
+  phys_strides[ndims() - 1] = 1;
+  for (int i = ndims() - 2; i >= 0; i--) {
+    phys_strides[i] = phys_strides[i + 1] * phys_dims[i + 1];
+  }
+  return phys_strides;
+}
+
+std::vector<int64_t> TensorDescriptor::GetLogicalStrides() const {
+  std::vector<int64_t> physical_strides = GetPhysicalStridesMajorToMinor();
+  std::reverse(physical_strides.begin(), physical_strides.end());
```
Comment (akuegel) [inline tensorflow/compiler/xla/stream_executor/dnn.cc:583]: This requires ```#include <algorithm>``` at the top of the file.

File: tensorflow/compiler/xla/stream_executor/dnn.h [hunk 6]
```diff
@@ -152,6 +152,75 @@ enum class RnnDirectionMode {
 // DepthToSpace comment for more information.
 enum class DepthToSpaceLayout { DepthHeightWidth };
 
+class TensorDescriptor {
+ public:
+  TensorDescriptor() {}
```
Comment (akuegel) [inline tensorflow/compiler/xla/stream_executor/dnn.h:157]: Please use TensorDescriptor() = default;

File: tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.h [hunk 7]
```diff
@@ -276,6 +276,54 @@ class CudnnSupport : public dnn::DnnSupport {
...
+  tsl::StatusOr<std::unique_ptr<const dnn::FusedMHASoftmaxRunner>>
+  FusedMHASoftmaxRunnerFromDesc(
+      Stream* stream, const dnn::AlgorithmDesc& algorithm_desc,
+      dnn::FusedMHAKind kind,
+      const dnn::MatmulTensorDescriptor& bmm1_lhs_descriptor,
+      const dnn::MatmulTensorDescriptor& bmm1_rhs_descriptor,
+      const dnn::MatmulTensorDescriptor& bmm2_rhs_descriptor,
+      const dnn::MatmulTensorDescriptor& intermediate_bmm2_lhs_descriptor,
+      const dnn::TensorDescriptor& output_descriptor,
+      std::optional<double> dropout_rate, std::optional<int64_t> seed);
```
Comment (akuegel) [inline tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.h:288]: missing "override" at the end.

**Review by akuegel — CHANGES_REQUESTED** [review]
I attached comments to the places I had to adjust because of ClangTidy and Linter.
It might be easier for you if you revert the revert PR, then you have all those fixes already and can add additional fixes for the build time issues on top.
